design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+  }
+}
+```
+
+We can always add additional configuration options later. We may also wish to expose the Vert.x `HttpClientOptions` for
+advanced users that wish to provide custom configs.
+
+### Transient queries -- Streaming
+
+The `Client` interface will provide the following methods for streaming the results of a transient (push or pull) query:
+```
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive the results one row at a time.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<QueryResult> streamQuery(String sql, Map<String, Object> properties);
+```
+where `QueryResult` is as follows:
+```
+/**
+ * The result of a query (push or pull), streamed one row at time. Records may be consumed by either
+ * subscribing to the publisher or polling (blocking) for one record at a time. These two methods of
+ * consumption are mutually exclusive; only one method may be used (per QueryResult).
+ */
+public interface QueryResult extends Publisher<Row> {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  String queryID();
+
+  /**
+   * Block until a row becomes available.
+   *
+   * @return the row.
+   */
+  Row poll();
+
+  /**
+   * Block until a row becomes available or the timeout has elapsed.
+   *
+   * @param timeout amount of to wait for a row. Non-positive values are interpreted as no timeout.
+   * @param timeUnit unit for timeout param.
+   * @return the row, if available; else, null.
+   */
+  Row poll(long timeout, TimeUnit timeUnit);
+
+  boolean isComplete();
+
+  void close();
+}
+```
+Note that `QueryResult` is a Reactive Streams `Publisher` so users can stream results. Users can also call `poll()` to receive
+results in a synchronous fashion instead. Only one of the two methods will be allowed per `QueryResult` instance.
+
+The `Row` interface is as follows:
+```
+/**
+ * A single record, returned as part of a query result.
+ */
+public interface Row {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  List<Object> values();
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Object getObject(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Object getObject(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  String getString(int columnIndex);
```

Comment:
what happens if you call any of these methods on a column that isn't of the specified type?

Reply:
We should throw a client exception. I haven't thought too much about what that interface should look like. I assume a standard error message + cause should be enough?

Reply:
I don't have strong thoughts on what the cause in this case should be. Suggestions welcome.

Reply:
yeah probably fine, I never like the `getX` pattern but I don't think there's a better one for this use case :(

Reply:
Probably the "right" exception to be thrown here is a ClassCastException, but most likely they throw that anyway, but probably best to check that with a test.

Reply:
I think, exceptions like this make sense and should be documented in the KLIP (of course, also later in the JavaDocs).